node-graph/gcore/src/vector/merge_by_distance.rs

```diff
@@ -0,0 +1,109 @@
+//! Collapse all nodes with all edges < distance
+
+use glam::DVec2;
+use petgraph::prelude::UnGraphMap;
+use rustc_hash::FxHashSet;
+
+use super::PointId;
+use super::VectorData;
+
+use super::VectorDataIndex;
+
+impl VectorData {
+	/// Collapse all nodes with all edges < distance
+	pub(crate) fn merge_by_distance(&mut self, distance: f64) {
+		// treat self as an undirected graph with point = node, and segment = edge
+		let indices = VectorDataIndex::build_from(self);
+
+		// Graph that will contain only short edges. References data graph
+		let mut short_edges = UnGraphMap::new();
+		for seg_id in self.segment_ids().iter().copied() {
+			let length = indices.segment_chord_length(seg_id);
+			if length < distance {
+				let [start, end] = indices.segment_ends(seg_id);
+				let start = indices.point_graph.node_weight(start).unwrap().id;
+				let end = indices.point_graph.node_weight(end).unwrap().id;
+
+				short_edges.add_node(start);
+				short_edges.add_node(end);
+				short_edges.add_edge(start, end, seg_id);
+			}
+		}
+
+		// Now group connected segments - all will be collapsed to a single point.
+		// Note: there are a few algos for this - perhaps test empirically to find fastest
+		let collapse: Vec<FxHashSet<PointId>> = petgraph::algo::tarjan_scc(&short_edges).into_iter().map(|connected| connected.into_iter().collect()).collect();
```

Comment:
Tarjan will almost certainly be faster than Kosaraju because it only requires a single graph traversal